### PR TITLE
En_Torch2

### PR DESCRIPTION
**Pull request by petrie911**
As the name suggests, this is Dark Link. I'm certain that Update is equivalent, but I can't get the static variables to load from pointer vs direct properly.

### CODE REVIEW

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c [hunk 1]
```diff
@@ -10,12 +10,51 @@
 
 #define THIS ((Player*)thisx)
 
+typedef enum {
```
Comment (Kenix3) [inline src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:13]: Should we rename this to dark link

**Review by Kenix3 — COMMENTED** [review]
Looks good to, but want to start a conversation on whether to rename things like entorch2 enums to dark link.

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c (hunk 1 above)
Comment (NEstelami) [reply src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:13]: IIRC in the long term we would be renaming everything to use names that made more sense, but I think that's for later.

Comment (fig02) [reply src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:13]: I personally do use the better name for actors in things like temp variables and enums, but in the end its stuff that will be discussed and enforced at a future documentation stage anyway, so its left to the discretion of the person decopmiling for now

File: src/overlays/actors/ovl_En_Blkobj/z_en_blkobj.c [hunk 2]
```diff
@@ -53,95 +53,95 @@ extern Gfx D_060014E0[];
...
+            this->alpha = 255;
+            EnBlkobj_SetupAction(this, EnBlkobj_Noop);
+            DynaPolyInfo_Free(globalCtx, &globalCtx->colCtx.dyna, this->dyna.dynaPolyId);
         }
     }
 }
 
-void func_809C22F4(EnBlkobj* this, GlobalContext* globalCtx) {
+void EnBlkobj_Noop(EnBlkobj* this, GlobalContext* globalCtx) {
 }
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Blkobj/z_en_blkobj.c:127]: id go with DoNothing in this case since it is an action

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c [hunk 3]
```diff
@@ -10,12 +10,51 @@
...
+    /* 19 */ JUMPSLASH_FINISH,
+    /* 20 */ BACKSLASH_RIGHT,
+    /* 21 */ BACKSLASH_LEFT,
+    /* 22 */ HAMMER_FORWARD,
+    /* 23 */ HAMMER_SIDE,
+    /* 24 */ SPIN_ATTACK_1H,
+    /* 25 */ SPIN_ATTACK_2H,
+    /* 26 */ BIG_SPIN_1H,
+    /* 27 */ BIG_SPIN_2H
+} PlayerSwordAnimation;
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:49]: def would be pretty weird to have this enum for dark link and not use it with player as well.
@Roman971 not sure if you want to wait for a player doc pass, but yeah just my opinion

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c [hunk 4]
```diff
@@ -27,21 +66,715 @@ const ActorInit En_Torch2_InitVars = {
...
+/* static */ u8 sStaggerCount;
+/* static */ u8 sStaggerTimer;
+/* static */ s8 sLastSwordAnim;
+/* static */ u8 sAlpha;
+
+void EnTorch2_Init(Actor* thisx, GlobalContext* globalCtx2) {
+    static DamageTable damageTable = {
+        0x10, 0x02, 0x01, 0x02, 0x10, 0x02, 0x02, 0x10, 0x01, 0x02, 0x04, 0x02, 0x02, 0x02, 0x02, 0x02,
+        0x02, 0xE2, 0x60, 0xD3, 0x00, 0x00, 0x01, 0x04, 0x02, 0x02, 0x08, 0x04, 0x00, 0x00, 0x04, 0x00,
+    };
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:95]: bring this out in front of the function

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c [hunk 5]
```diff
@@ -27,21 +66,715 @@ const ActorInit En_Torch2_InitVars = {
...
+    this->swordQuads[0].base.acFlags = this->swordQuads[1].base.acFlags = 0xD;
+    this->swordQuads[0].base.type = this->swordQuads[1].base.type = 9;
+    this->swordQuads[0].body.toucher.damage = this->swordQuads[1].body.toucher.damage = 8;
+    this->swordQuads[0].body.bumperFlags = this->swordQuads[1].body.bumperFlags = 1;
+    this->shieldQuad.base.atFlags = 0x11;
+    this->shieldQuad.base.acFlags = 0xD;
+    this->actor.colChkInfo.damageTable = &damageTable;
+    this->actor.colChkInfo.health = gSaveContext.healthCapacity >> 3;
+    this->actor.colChkInfo.unk_10 = 0x3C;
+    this->actor.colChkInfo.unk_12 = 0x64;
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:117]: can make the cylinder dimensions decimal if you want

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c [hunk 6]
```diff
@@ -27,21 +66,715 @@ const ActorInit En_Torch2_InitVars = {
...
+
+#ifdef NON_MATCHING
+void EnTorch2_Update(Actor* thisx, GlobalContext* globalCtx2) {
+    GlobalContext* globalCtx = globalCtx2;
+    Player* player = PLAYER;
+    Player* this = THIS;
+    Input* input = &sInput;
+    u16 phi_a2;
+    s8 tempx;
+    s8 tempy;
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:209]: tempX tempY

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c [hunk 7]
```diff
@@ -27,21 +66,715 @@ const ActorInit En_Torch2_InitVars = {
...
+    u32 phi_v0;
+    Actor* attackItem;
+    s16 sp5A;
+    s16 pad58;
+    s32 pad54;
+    f32 sp50;
+    s16 sp4E;
+    s16 pad4C;
+
+    sp5A = player->actor.shape.rot.y - this->actor.shape.rot.y;
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:223]: could name this yawDiff if you want

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c [hunk 8]
```diff
@@ -27,21 +66,715 @@ const ActorInit En_Torch2_InitVars = {
...
+                            EnTorch2_SwingSword(globalCtx, input, this);
+                            sSwordJumpState++;
+                        } else if (sSwordJumpTimer == 19) {
+                            func_800F4190(&this->actor.projectedPos, NA_SE_VO_LI_AUTO_JUMP);
+                        }
+                    }
+                } else {
+                    /*
+                     * This does nothing, as sHoldShieldTimer is never set.
+                     */
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:350]: this can be a // comment to take up less space

imo all of these can be single line comments, but i understand wanting to make the big ones stand out to break up this very long function. i leave that up to you

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c [hunk 9]
```diff
@@ -27,21 +66,715 @@ const ActorInit En_Torch2_InitVars = {
...
+                                sStickAngle += -0x4000; // Not loaded from pointer
+                            } else if (player->swordAnimation <= HAMMER_SIDE) {
+                                input->cur.button = BTN_R;
+                            } else if (player->swordAnimation <= BIG_SPIN_2H) {
+                                EnTorch2_Backflip(this, input, &this->actor);
+                            } else {
+                                EnTorch2_Backflip(this, input, &this->actor);
+                            }
+                            if (!(input->cur.button & (BTN_A | BTN_R)) && !this->swordState && player->swordState) {
+                                sCounterState = 1;
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:409]: is sword state a bool? had the impression from earlier that it wasnt.
if it is then itll have to be made consistent

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c [hunk 10]
```diff
@@ -27,21 +66,715 @@ const ActorInit En_Torch2_InitVars = {
...
+                            }
+                        }
+                    } else {
+                        /*
+                         * Handles movement and attacks when not reacting to Link's actions
+                         */
+                        sStickAngle = this->actor.yawTowardsLink;
+                        sp50 = 0.0f;
+                        if ((this->actor.xzDistFromLink <= 90.0f) && (70.0f < this->actor.xzDistFromLink) &&
+                            (ABS(sp5A) >= 0x7800) &&
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:419]: id put 70 on the right side of compare if possible

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c [hunk 11]
```diff
@@ -27,21 +66,715 @@ const ActorInit En_Torch2_InitVars = {
...
+                        } else if (this->actor.xzDistFromLink <= (50 + sp50)) {
+                            sStickTilt = 127.0f;
+                            sStickAngle = this->actor.yawTowardsLink;
+                            if (this->actor.unk_10C == 0) {
+                                Math_SmoothScaleMaxMinS(&sStickAngle, player->actor.shape.rot.y + 0x7FFF, 1, 0x2328, 0);
+                            }
+                        } else if ((100.0f + sp50) < this->actor.xzDistFromLink) {
+                            if (!player->swordState || (player->swordAnimation < SPIN_ATTACK_1H) ||
+                                (player->swordAnimation > BIG_SPIN_2H) || (280.0f <= this->actor.xzDistFromLink)) {
+                                sStickTilt = 127.0f;
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:438]: 280 on right if possible

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c [hunk 12]
```diff
@@ -27,21 +66,715 @@ const ActorInit En_Torch2_InitVars = {
...
+    } else if (sSwordJumpHeight != 0) {
+        this->actor.posRot.pos.y += sSwordJumpHeight * 0.01f;
+        sSwordJumpHeight = 0;
+    }
+    if ((sActionState == ENTORCH2_WAIT) || (this->invincibilityTimer < 0)) {
+        sZTargetFlag = false;
+    } else {
+        sZTargetFlag = true;
+    }
+    DECR(sJumpslashTimer);
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:732]: decr

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c [hunk 13]
```diff
@@ -27,21 +66,715 @@ const ActorInit En_Torch2_InitVars = {
...
+                                               EnTorch2_PostLimbDraw, this, POLY_OPA_DISP);
+    } else {
+        gDPSetEnvColor(POLY_XLU_DISP++, 255, 0, 0, sAlpha);
+        gSPSegment(POLY_XLU_DISP++, 0x0C, D_80116280);
+        func_8002EBCC(&this->actor, globalCtx, 0);
+        func_8002ED80(&this->actor, globalCtx, 0);
+        POLY_XLU_DISP = SkelAnime_DrawFlex(globalCtx, this->skelAnime.skeleton, this->skelAnime.limbDrawTbl,
+                                               this->skelAnime.dListCount, EnTorch2_OverrideLimbDraw,
+                                               EnTorch2_PostLimbDraw, this, POLY_XLU_DISP);
+    }
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:778]: run formatter

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c (hunk 8 above)
Comment (petrie911) [reply src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:350]: Yeah, that was definitely intentional. The if tree is very hard to follow without them.

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c (hunk 9 above)
Comment (petrie911) [reply src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:409]: It is not, even though it's essentially use that way here, because there's a slight distinction between `-1` and `1`. Incidentally, other actors with swords (like EnGeldB) also do this.

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c (hunk 3 above)
Comment (petrie911) [reply src/overlays/actors/ovl_En_Torch2/z_en_torch2.c]: It's kind of here as a plant for Player to use it. And there's little reason not to (it's definitely correct) aside from the fact that implementing it into z_player would likely take a bit of time.

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c (hunk 9 above)
Comment (fig02) [reply src/overlays/actors/ovl_En_Torch2/z_en_torch2.c]: ok, then yeah i believe there are some comparisons like `if (!this->swordState)` that need to be changed to non bool

Comment (petrie911) [reply src/overlays/actors/ovl_En_Torch2/z_en_torch2.c]: and they have been.

Comment (fig02) [reply src/overlays/actors/ovl_En_Torch2/z_en_torch2.c]: there are still some left that need to change

Comment (petrie911) [reply src/overlays/actors/ovl_En_Torch2/z_en_torch2.c]: There are? I thought I got them all.

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c [hunk 14]
```diff
@@ -27,21 +66,720 @@ const ActorInit En_Torch2_InitVars = {
...
+            } else if (attackItem != NULL) {
+                sDodgeRollState = 1;
+                sStickAngle = this->actor.yawTowardsLink;
+                sStickTilt = 127.0f;
+                input->cur.button = BTN_A;
+            } else if (sJumpslashTimer == 0) {
+                /*
+                 * Handles Dark Link's initial reaction to jumpslashes
+                 */
+                if ((player->swordState || (player->actor.velocity.y > -3.0f)) &&
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:307]: swordState

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c [hunk 15]
```diff
@@ -27,21 +66,720 @@ const ActorInit En_Torch2_InitVars = {
...
+                                sStickAngle += -0x4000; // Not loaded from pointer
+                            } else if (player->swordAnimation <= HAMMER_SIDE) {
+                                input->cur.button = BTN_R;
+                            } else if (player->swordAnimation <= BIG_SPIN_2H) {
+                                EnTorch2_Backflip(this, input, &this->actor);
+                            } else {
+                                EnTorch2_Backflip(this, input, &this->actor);
+                            }
+                            if (CHECK_BTN_ANY(input->cur.button, BTN_A | BTN_R) && (this->swordState == 0) &&
+                                player->swordState) {
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:412]: swordState

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c [hunk 16]
```diff
@@ -27,21 +66,720 @@ const ActorInit En_Torch2_InitVars = {
...
+                         */
+                        sStickAngle = this->actor.yawTowardsLink;
+                        sp50 = 0.0f;
+                        if ((90.0f >= this->actor.xzDistFromLink) && (this->actor.xzDistFromLink > 70.0f) &&
+                            (ABS(sp5A) >= 0x7800) &&
+                            ((this->actor.unk_10C != 0) || !(player->stateFlags1 & 0x00400000))) {
+                            EnTorch2_SwingSword(globalCtx, input, this);
+                        } else if (((this->actor.xzDistFromLink <= 70.0f) ||
+                                    ((this->actor.xzDistFromLink <= 80.0f + sp50) && player->swordState)) &&
+                                   (this->swordState == 0)) {
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:428]: swordState

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c [hunk 17]
```diff
@@ -27,21 +66,720 @@ const ActorInit En_Torch2_InitVars = {
...
+                            }
+                        } else if (this->actor.xzDistFromLink <= (50 + sp50)) {
+                            sStickTilt = 127.0f;
+                            sStickAngle = this->actor.yawTowardsLink;
+                            if (this->actor.unk_10C == 0) {
+                                Math_SmoothScaleMaxMinS(&sStickAngle, player->actor.shape.rot.y + 0x7FFF, 1, 0x2328, 0);
+                            }
+                        } else if ((100.0f + sp50) < this->actor.xzDistFromLink) {
+                            if (!player->swordState || (player->swordAnimation < SPIN_ATTACK_1H) ||
+                                (player->swordAnimation > BIG_SPIN_2H) || (this->actor.xzDistFromLink >= 280.0f)) {
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:441]: swordState

Comment (petrie911) [reply src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:441]: oh, lol, searched for this->swordState. Forgot there's also player->swordState.

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c (hunk 8 above)
Comment (fig02) [reply src/overlays/actors/ovl_En_Torch2/z_en_torch2.c]: i still think a // comment will be just as helpful and be a bit less clutter-ey but its not the biggest deal. can see what others think too

**Review by fig02 — APPROVED** [review]

File: src/overlays/actors/ovl_En_Blkobj/z_en_blkobj.c [hunk 18]
```diff
@@ -39,12 +39,12 @@ static InitChainEntry sInitChain[] = {
     ICHAIN_F32(uncullZoneDownward, 300, ICHAIN_STOP),
 };
 
-static Gfx D_809C2590[] = {
+static Gfx sSetOpa[] = {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Blkobj/z_en_blkobj.c:42]: For this kind of display lists, I suggest something a bit more descriptive like `sSetupOpaDL`

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c [hunk 19]
```diff
@@ -27,21 +66,720 @@ const ActorInit En_Torch2_InitVars = {
...
+/* static */ u8 sActionState;
+/* static */ u8 sSwordJumpTimer;
+/* static */ u8 sCounterState;
+/* static */ u8 sDodgeRollState;
+/* static */ u8 sStaggerCount;
+/* static */ u8 sStaggerTimer;
+/* static */ s8 sLastSwordAnim;
+/* static */ u8 sAlpha;
+
+static DamageTable damageTable = {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Torch2/z_en_torch2.c:91]:
```suggestion
static DamageTable sDamageTable = {
```

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c (hunk 3 above)
Comment (Roman971) [reply src/overlays/actors/ovl_En_Torch2/z_en_torch2.c]: It's definitely going to be weird until we have unified everything, but since it's already there and used in dark link, I think it's fine to leave at that for the time being.

File: src/overlays/actors/ovl_En_Torch2/z_en_torch2.c (hunk 8 above)
Comment (Roman971) [reply src/overlays/actors/ovl_En_Torch2/z_en_torch2.c]: I agree with Fig here. In general, I prefer leaving multi line comments to function/file level documentation, so it's easier to distinguish them when going through the file. 

Also, as a small related note, the style for multi line comments is to start with `/**` on the first line, and always break before and after the comment's body. VSCode (like most IDEs) will even "understand" it's a multi line comment if you start it with `/**` then press enter to start typing.

Comment (Roman971) [reply src/overlays/actors/ovl_En_Torch2/z_en_torch2.c]: Feel free to leave more spaces between blocks if you need to tho. I don't mention it in reviews, but nowadays I think a lot of clarity could be gained in the codebase by leaving empty lines in the right places. For example here I would leave an empty line after each switch case and after each if block, at the very least.

**Review by Roman971 — APPROVED** [review]
